moveit_ros/moveit_servo/CMakeLists.txt

```
@@ -85,7 +85,7 @@ set_target_properties(${SERVO_LIB_NAME} PROPERTIES VERSION "${${PROJECT_NAME}_VE
 ament_target_dependencies(${SERVO_LIB_NAME} ${THIS_PACKAGE_INCLUDE_DEPENDS})
 target_link_libraries(${SERVO_LIB_NAME} ${SERVO_PARAM_LIB_NAME})
 
-add_library(${POSE_TRACKING} SHARED src/pose_tracking.cpp)
+add_library(${POSE_TRACKING} SHARED src/pose_tracking.cpp src/servo_parameters.cpp)
 ament_target_dependencies(${POSE_TRACKING} ${THIS_PACKAGE_INCLUDE_DEPENDS})
 target_link_libraries(${POSE_TRACKING} ${SERVO_LIB_NAME})
```

I think we should link it against the already built library rather than adding the source files to the target

```suggestion
add_library(${POSE_TRACKING} SHARED src/pose_tracking.cpp)
ament_target_dependencies(${POSE_TRACKING} ${THIS_PACKAGE_INCLUDE_DEPENDS})
target_link_libraries(${POSE_TRACKING} ${SERVO_LIB_NAME} ${SERVO_PARAM_LIB_NAME})
```

---

You can't link against the already-built library, because as you can see from the failing CI, this is a template issue -- the already-built library doesn't have the correct templated classes built in, which is why linker fails. Thus, the source has to be built each time (as is done here) to allow for any additional template classes that need to be built.

Failing CI of this change: https://github.com/nkalupahana/ros2-foxy-macos/runs/3058543555?check_suite_focus=true

---

> I think including the implementation (cpp) in the .h file would also work; that's the pattern I've seen for template classes in the past, at least. Let me know what you want to see.

Yes, I think this should be the solution

---

Alright, so it looks like you can't just include the .cpp file because not everything in there is templated so if you include it, then you'll get duplicate symbols. To fix that, I just moved the template function into the .h file (the forward declaration is already in there, so I figured that wouldn't be too big of a deal). Let me know if this works!

CI: https://github.com/nkalupahana/ros2-foxy-macos/actions/runs/1027725851

---

I'd like to recommend the rationale for this be added as a comment to this CMake file... just copy-paste your explanation here, into the code :-)

---

Added! :)